src/main/java/org/owasp/webgoat/lessons/sqlinjection/advanced/SqlInjectionChallengeBasic.java

```diff
@@ -0,0 +1,105 @@
+/*
+ * This file is part of WebGoat, an Open Web Application Security Project utility. For details, please see http://www.owasp.org/
+ *
+ * Copyright (c) 2002 - 2019 Bruce Mayhew
+ *
+ * This program is free software; you can redistribute it and/or modify it under the terms of the
+ * GNU General Public License as published by the Free Software Foundation; either version 2 of the
+ * License, or (at your option) any later version.
+ *
+ * This program is distributed in the hope that it will be useful, but WITHOUT ANY WARRANTY; without
+ * even the implied warranty of MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE. See the GNU
+ * General Public License for more details.
+ *
+ * You should have received a copy of the GNU General Public License along with this program; if
+ * not, write to the Free Software Foundation, Inc., 59 Temple Place - Suite 330, Boston, MA
+ * 02111-1307, USA.
+ *
+ * Getting Source ==============
+ *
+ * Source for this application is maintained at https://github.com/WebGoat/WebGoat, a repository for free software projects.
+ */
+
+ package org.owasp.webgoat.lessons.sqlinjection.advanced;
+
+ import java.sql.*;
+ import lombok.extern.slf4j.Slf4j;
+ import org.owasp.webgoat.container.LessonDataSource;
+ import org.owasp.webgoat.container.assignments.AssignmentEndpoint;
+ import org.owasp.webgoat.container.assignments.AssignmentHints;
+ import org.owasp.webgoat.container.assignments.AttackResult;
+ import org.springframework.util.StringUtils;
+ import org.springframework.web.bind.annotation.PutMapping;
+ import org.springframework.web.bind.annotation.RequestParam;
+ import org.springframework.web.bind.annotation.ResponseBody;
+ import org.springframework.web.bind.annotation.RestController;
+ 
+ /**
+  * @author nbaars
+  * @since 4/8/17.
+  */
+ @RestController
+ @AssignmentHints(
+     value = {"SqlInjectionChallengeBasic4", "SqlInjectionChallengeBasic5", "SqlInjectionChallengeBasic6"})
+ @Slf4j
+ public class SqlInjectionChallengeBasic extends AssignmentEndpoint {
+ 
+   private final LessonDataSource dataSource;
+ 
+   public SqlInjectionChallengeBasic(LessonDataSource dataSource) {
+     this.dataSource = dataSource;
+   }
+ 
+   @PutMapping("/SqlInjectionAdvanced/challengeNew")
+   // assignment path is bounded to class so we use different http method :-)
+   @ResponseBody
+   public AttackResult registerNewUser(
+       @RequestParam String username_reg,
+       @RequestParam String email_reg,
+       @RequestParam String password_reg)
+       throws Exception {
+     AttackResult attackResult = checkArguments(username_reg, email_reg, password_reg);
+ 
+     if (attackResult == null) {
+ 
+       try (Connection connection = dataSource.getConnection()) {
+         String checkUserQuery =
+             "select userid from sql_challenge_users where userid = '" + username_reg + "'";
+         Statement statement = connection.createStatement();
+         ResultSet resultSet = statement.executeQuery(checkUserQuery);
```

Comment:
Hey @aegilops 👋 Pulling you in as a reviewer here. Can you walk me through how I may fix this please? I have read the alert information [here](https://github.com/octodemo/forrester-webgoat/security/code-scanning/77), but still am a little confused. I really want to fix this myself, but could need some help, if at all possible. Any explanation would be great 🙇

Reply:
No problem Nick!

The problem is that the endpoint you added takes user input from `username_reg` and uses it to create a SQL query in an unsafe way 💥. It concatenates `+` the string with a SQL query, instead of using a parameterized query with `?` placeholders.

If you change it to be parameterized instead it'll be secure against SQL injection (the name for this kind of vulnerability).

If you're not sure what to do, have a go at using CoPilot Labs' "make robust" brush 🛡️, or if you have access to the private beta of CoPilot Chat 💬 you can do it in there too, just by asking it to fix up the code you've highlighted.

Jump on a Teams call with me and I can walk you through doing that in Codespaces, if you like?

Reply:
Thanks @aegilops that makes sense! I have assigned this to myself to fix 🙇 Which can be found here: https://github.com/octodemo/forrester-webgoat/issues/4 :)